Support generics in Mutable StateVars

If I use generics in Mutable StateVars:

```
      @StateVar(value = "hoplaa", readOnly = true) Mutable<Map<Integer, Integer>> hoplaa,
```

I get the following error in startup:

```
Caused by: java.lang.ClassCastException: class sun.reflect.generics.reflectiveObjects.ParameterizedTypeImpl cannot be cast to class java.lang.Class (sun.reflect.generics.reflectiveObjects.ParameterizedTypeImpl and java.lang.Class are in module java.base of loader 'bootstrap')
	at io.nflow.engine.internal.workflow.WorkflowDefinitionScanner.lambda$getStateMethods$0(WorkflowDefinitionScanner.java:72) ~[nflow-engine-7.4.0.jar:na]
	at org.springframework.util.ReflectionUtils.doWithMethods(ReflectionUtils.java:367) ~[spring-core-5.3.20.jar:5.3.20]
	at io.nflow.engine.internal.workflow.WorkflowDefinitionScanner.getStateMethods(WorkflowDefinitionScanner.java:56) ~[nflow-engine-7.4.0.jar:na]
	at io.nflow.engine.workflow.definition.AbstractWorkflowDefinition.<init>(AbstractWorkflowDefinition.java:55) ~[nflow-engine-7.4.0.jar:na]
	at io.nflow.engine.workflow.definition.AbstractWorkflowDefinition.<init>(AbstractWorkflowDefinition.java:40) ~[nflow-engine-7.4.0.jar:na]
	at io.nflow.engine.workflow.definition.WorkflowDefinition.<init>(WorkflowDefinition.java:37) ~[nflow-engine-7.4.0.jar:na]
	at fi.mybusiness.workflow.MyWorkflow.<init>(MyWorkflow.java:105) ~[main/:na]
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[na:na]
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490) ~[na:na]
	at org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:211) ~[spring-beans-5.3.20.jar:5.3.20]
	... 65 common frames omitted
```

Observed using OpenJDK 11.0.15.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -39,6 +39,9 @@
   - Workflow instance actions and state variables that are older than 45 days are automatically cleaned up occasionally when the instance is processed. Use workflow settings to change the default time period (`setHistoryDeletableAfter`) and condition (`setDeleteHistoryCondition`) of the clean-up.
   - Add support to query also archived workflow instances when not enough non-archived matches are found.
   - Add `WorkflowInstance.Builder.setState(WorkflowState)` convenience method.
+  - Detect state methods with duplicate @StateVar names
+  - Allow simple generics objects in Mutable state variables
+  - Support instantiating Collection,List and Map interfaces to empty in state variables
   - Dependency updates
     - logback-classic update to version 1.2.11
       - <http://mailman.qos.ch/pipermail/announce/2021/000164.html>

--- a/nflow-engine/src/main/java/io/nflow/engine/internal/workflow/WorkflowDefinitionScanner.java
+++ b/nflow-engine/src/main/java/io/nflow/engine/internal/workflow/WorkflowDefinitionScanner.java
@@ -20,7 +20,9 @@ import java.lang.reflect.Type;
 import java.math.BigDecimal;
 import java.math.BigInteger;
 import java.util.ArrayList;
+import java.util.Collection;
 import java.util.Collections;
+import java.util.HashMap;
 import java.util.HashSet;
 import java.util.LinkedHashMap;
 import java.util.LinkedHashSet;
@@ -74,6 +76,9 @@ public class WorkflowDefinitionScanner {
             if (Mutable.class.isAssignableFrom(clazz)) {
               ParameterizedType pType = (ParameterizedType) type;
               type = pType.getActualTypeArguments()[0];
+              if (type instanceof ParameterizedType) {
+                type = ((ParameterizedType) type).getRawType();
+              }
               clazz = (Class<?>) type;
               readOnly = false;
               mutable = true;
@@ -88,12 +93,22 @@ public class WorkflowDefinitionScanner {
         throw new IllegalStateException(
             "Not all parameter names could be resolved for " + method + ". Maybe missing @StateVar annotation?");
       }
+      assertNoDuplicates(method, params);
       if (methods.containsKey(method.getName())) {
         throw new IllegalStateException("Method " + method + " was overloaded. Overloading state methods is not allowed.");
       }
       methods.put(method.getName(), new WorkflowStateMethod(method, params.toArray(new StateParameter[params.size()])));
     }, new WorkflowTransitionMethod());
     return methods;
+  }
+
+  private void assertNoDuplicates(Method method, Collection<StateParameter> params) {
+    Set<String> names = new HashSet<>(params.size(), 1);
+    for (StateParameter p : params) {
+      if (!names.add(p.key)) {
+        throw new IllegalStateException("Method " + method + " had to state variables with same name: " + p.key + ".");
+      }
+    }
   }
 
   public Set<WorkflowState> getPublicStaticWorkflowStates(Class<? extends WorkflowDefinition> definition) {
@@ -123,6 +138,13 @@ public class WorkflowDefinitionScanner {
       return invokeMethod(findMethod(primitiveToWrapper(clazz), "valueOf", String.class), null, "0");
     }
     if (stateInfo != null && stateInfo.instantiateIfNotExists()) {
+      if (clazz.isInterface()) {
+        if (Map.class.isAssignableFrom((clazz))) {
+          clazz = HashMap.class;
+        } else if (Collections.class.isAssignableFrom((clazz)) || List.class.isAssignableFrom((clazz))) {
+          clazz = ArrayList.class;
+        }
+      }
       try {
         Constructor<?> ctr = clazz.getConstructor();
         ctr.newInstance();

--- a/nflow-engine/src/test/java/io/nflow/engine/internal/executor/WorkflowStateProcessorTest.java
+++ b/nflow-engine/src/test/java/io/nflow/engine/internal/executor/WorkflowStateProcessorTest.java
@@ -665,6 +665,7 @@ public class WorkflowStateProcessorTest extends BaseNflowTest {
     assertThat(state.get("nullPojo"), is("{\"field\":\"magical instance\",\"test\":false}"));
     assertThat(state.get("immutablePojo"), is("{\"field\": \"unmodified\"}"));
     assertThat(state.get("mutableString"), is("mutated"));
+    assertThat(state.get("genericsMutable"), is("{\"mutated\":\"true\"}"));
     assertThat(state.get("hello"), is("[1,2,3]"));
   }
 
@@ -1023,17 +1024,19 @@ public class WorkflowStateProcessorTest extends BaseNflowTest {
     public NextAction process(StateExecution execution, @StateVar("string") String s, @StateVar("int") int i,
         @StateVar("pojo") Pojo pojo, @StateVar(value = "nullPojo", instantiateIfNotExists = true) Pojo pojo2,
         @StateVar(value = "immutablePojo", readOnly = true) Pojo unmodifiablePojo, @StateVar("nullInt") int zero,
-        @StateVar("mutableString") Mutable<String> mutableString) {
+        @StateVar("mutableString") Mutable<String> mutableString,
+        @StateVar(value="genericsMutable", instantiateIfNotExists = true) Mutable<Map<String, String>> genericsMutable) {
       assertThat(execution.getWorkflowInstanceId(), is(1L));
       assertThat(execution.getWorkflowInstanceExternalId(), is(notNullValue()));
       Pojo pojo1 = execution.getVariable("pojo", Pojo.class);
       assertThat(pojo.field, is(pojo1.field));
       assertThat(pojo.test, is(pojo1.test));
-      lastArgs = asList(s, i, pojo, pojo2, unmodifiablePojo, zero, mutableString.val);
+      lastArgs = asList(s, i, pojo, pojo2, unmodifiablePojo, zero, mutableString.val, genericsMutable.val);
       pojo.field += " modified";
       pojo2.field = "magical instance";
       unmodifiablePojo.field += " ignored";
       mutableString.val = "mutated";
+      genericsMutable.val.put("mutated", "true");
       execution.setVariable("hello", new int[] { 1, 2, 3 });
       return stopInState(TestState.DONE, "Finished");
     }

